File: backend/models/BloodUnit.js
import mongoose from "mongoose";

const bloodUnitSchema = new mongoose.Schema({
    barcode: { type: String, unique: true, required: true },
    donor: { type: mongoose.Schema.Types.ObjectId, ref: "Donor", required: true },
    donationDate: { type: Date, default: Date.now },
    bloodGroup: { type: String, required: true },
    productType: { type: String, enum: ["whole_blood", "red_cells", "platelets", "plasma"], default: "whole_blood" },
    quantity: { type: Number, default: 350 },
    status: { type: String, enum: ["quarantine", "available", "issued", "expired", "discarded"], default: "quarantine" },
    screeningStatus: { type: String, enum: ["pending", "passed", "failed"], default: "pending" },
    screening: {
        hiv: { type: String, enum: ["pending", "positive", "negative"], default: "pending" },
        hepatitisB: { type: String, enum: ["pending", "positive", "negative"], default: "pending" },
        hepatitisC: { type: String, enum: ["pending", "positive", "negative"], default: "pending" },
        syphilis: { type: String, enum: ["pending", "positive", "negative"], default: "pending" },
        malaria: { type: String, enum: ["pending", "positive", "negative"], default: "pending" },
        testedAt: Date,
        testedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" }
    },
    facility: { type: mongoose.Schema.Types.ObjectId, ref: "Facility" }, // nơi lưu trữ
    expiryDate: Date
}, { timestamps: true });

export default mongoose.model("BloodUnit", bloodUnitSchema);

File: backend-go/internal/models/blood_unit.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"
)

type BloodUnit struct {
	ID              primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	Barcode         string             `bson:"barcode" json:"barcode"`
	Donor           primitive.ObjectID `bson:"donor" json:"donor"`
	DonationDate    time.Time          `bson:"donationDate" json:"donationDate"`
	BloodGroup      string             `bson:"bloodGroup" json:"bloodGroup"`
	ProductType     string             `bson:"productType" json:"productType"` // whole_blood, red_cells, platelets, plasma
	Quantity        int                `bson:"quantity" json:"quantity"`
	Status          string             `bson:"status" json:"status"`                   // quarantine, available, issued, expired, discarded
	ScreeningStatus string             `bson:"screeningStatus" json:"screeningStatus"` // pending, passed, failed
	Screening       struct {
		HIV        string              `bson:"hiv" json:"hiv"`
		HepatitisB string              `bson:"hepatitisB" json:"hepatitisB"`
		HepatitisC string              `bson:"hepatitisC" json:"hepatitisC"`
		Syphilis   string              `bson:"syphilis" json:"syphilis"`
		Malaria    string              `bson:"malaria" json:"malaria"`
		TestedAt   *time.Time          `bson:"testedAt,omitempty" json:"testedAt,omitempty"`
		TestedBy   *primitive.ObjectID `bson:"testedBy,omitempty" json:"testedBy,omitempty"`
	} `bson:"screening" json:"screening"`
	Facility   *primitive.ObjectID `bson:"facility,omitempty" json:"facility,omitempty"`
	QRCode     string              `bson:"qrCode,omitempty" json:"qrCode,omitempty"`
	ExpiryDate *time.Time          `bson:"expiryDate,omitempty" json:"expiryDate,omitempty"`
	CreatedAt  time.Time           `bson:"createdAt" json:"createdAt"`
	UpdatedAt  time.Time           `bson:"updatedAt" json:"updatedAt"`
}
